core/src/main/kotlin/com/justai/jaicf/api/routing/BotRoutingEngine.kt

```diff
@@ -0,0 +1,151 @@
+        val executionContext = executionContext0 ?: ExecutionContext(requestContext, null, botContext, request)
+        setCurrentRoutingEngine(botContext, engineName)
+        botContext = applyNewDialogContext(botContext, engineName)
+
+        try {
+            logger.info("Process route request: ${request.input} with routing for engine: $engineName")
+            val engine = routables[engineName] ?: error("No engine with name: $engineName")
+            engine.process(request, reactions, requestContext, botContext, executionContext)
+        } catch (e: BotRequestRerouteException) {
+            botContext.routingContext.dialogContextMap[engineName] = botContext.dialogContext
+            processWithRouting(
+                request = request,
+                reactions = reactions,
+                requestContext = requestContext,
+                contextManager = contextManager,
+                botContext0 = botContext,
+                engineName = e.targetEngineName,
+                executionContext0 = executionContext
+            )
+        } finally {
+            botContext.routingContext.dialogContextMap[engineName] = botContext.dialogContext
+            (contextManager ?: main.defaultContextManager).saveContext(botContext,
+                request,
+                response = null,
+                requestContext)
+        }
+    }
+
+    private fun setCurrentRoutingEngine(botContext: BotContext, engineName: String) {
+        botContext.routingContext.currentEngine = engineName
+        try {
+            val curr = botContext.routingContext.routingEngineStack.peek()
+            if (curr != engineName) {
+                botContext.routingContext.routingEngineStack.push(engineName)
+            }
+        } catch (e: EmptyStackException) {
+            botContext.routingContext.routingEngineStack.push(engineName)
+        }
+    }
+
+    private fun applyNewDialogContext(ctx: BotContext, engineName: String): BotContext =
+        ctx.copy(dialogContext = ctx.routingContext.dialogContextMap.getOrDefault(engineName, DialogContext())).apply {
+            result = ctx.result
+            client.putAll(ctx.client)
+            session.putAll(ctx.session)
+            temp.putAll(ctx.temp)
+        }
+
+    override fun process(
+        request: BotRequest,
+        reactions: Reactions,
+        requestContext: RequestContext,
+        contextManager: BotContextManager?,
+    ) {
+        val botContext = (contextManager ?: main.defaultContextManager).loadContext(request, requestContext)
+        val route = getCurrentRouting(botContext) ?: DEFAULT_ROUTE_NAME
+        processWithRouting(request, reactions, requestContext, contextManager, botContext, route)
+    }
+
+    private fun getCurrentRouting(ctx: BotContext): String? = try {
+        ctx.routingContext.routingEngineStack.peek()
+    } catch (e: EmptyStackException) {
+        null
+    }
+
+    companion object {
+        internal const val DEFAULT_ROUTE_NAME = "main"
```

Comment:
Are you sure there is no way to get rid of hardcoded "main" engine name?
For example storing it with `null` key and making it possible to route to it with some explicit method like `routeToMainEngine`?

Reply:
I don't like to store it with null key, so i've changed default key to prevent conflicts on clients using name "main" in routables.